src/runtime/monitor/src/compartment.rs

```diff
@@ -0,0 +1,172 @@
+    thread_map: HashMap<ObjID, CompThreadInfo>,
+
+    name: String,
+}
+
+/// Safety: this is needed because of the comp_config field, but this points to an object that is
+/// same lifetime as Comp, and is unchanging.
+unsafe impl Send for Comp {}
+
+impl core::fmt::Debug for Comp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "comp:{}({:x}, {})",
+            &self.name, &self.sctx_id, &self.compartment_id
+        )
+    }
+}
+
+impl core::fmt::Display for Comp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", &self.name)
+    }
+}
+
+pub(crate) fn make_new_comp_alloc_object() -> ObjectHandle {
+    // TODO: in the future, we'll need to make this a runtime state object, and limit access rights.
+    let id = sys_object_create(
+        ObjectCreate::new(
+            BackingType::Normal,
+            LifetimeType::Volatile,
+            None,
+            ObjectCreateFlags::empty(),
+        ),
+        &[],
+        &[],
+    )
+    .unwrap();
+
+    twizzler_runtime_api::get_runtime()
+        .map_object(id.as_u128(), MapFlags::READ | MapFlags::WRITE)
+        .unwrap()
+}
+
+impl Comp {
+    /// Construct a new compartment. This can fail, since it allocates memory within the compartment
+    /// (for the TLS template and shared config data).
+    pub fn new<Backing: BackingData>(
+        sctx_id: ObjID,
+        compartment: &mut dynlink::compartment::Compartment<Backing>,
+    ) -> miette::Result<Self> {
+        // First make a new allocation object and initialize it.
+        let comp_alloc_obj = make_new_comp_alloc_object();
+        let mut allocator = Talc::new(ErrOnOom);
+        // Safety: the start and end pointers point within the same object.
+        unsafe {
+            let start = comp_alloc_obj.start.add(NULLPAGE_SIZE);
+            let end = start.add(MAX_SIZE - NULLPAGE_SIZE * 2);
+            // Unwrap-Ok: There is enough memory to claim, and the allocator is initialized.
+            allocator.claim(Span::new(start, end)).unwrap();
+        }
+
+        let mut comp = Self {
+            sctx_id,
+            compartment_id: compartment.id,
+            comp_alloc_obj,
+            thread_map: Default::default(),
+            allocator,
+            comp_config: OnceCell::new(),
+            name: compartment.name.clone(),
+        };
+
+        // Construct the TLS template.
+        let template_info = compartment
+            .build_tls_region(RuntimeThreadControl::new(0), |layout| {
+                comp.monitor_alloc(layout)
+            })?;
+
+        // Init the shared compartment config. We'll leak this TLS template since we are manually
+        // managing its lifetime.
+        let temp = Box::new(TlsTemplateInfo::from(template_info));
+        let temp = Box::leak(temp);
+        let cc = comp
+            .monitor_new(SharedCompConfig::new(sctx_id, temp))
```

Comment:
So, the box::new here is wrong, but the addresses in the tls template get rebased over to the compartment when it loads the template into a new tls region for a new thread. See TlsTemplateInfo::init_new_tls_region.

I have fixed the Box::new issue in a future upcoming patch, since it doesn't matter yet in this PR (none of it is secure anyway)